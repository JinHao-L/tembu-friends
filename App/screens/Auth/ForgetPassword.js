import React, { Component } from 'react';
import { View, StyleSheet, TouchableWithoutFeedback, Keyboard, Platform } from 'react-native';

import { withFirebase } from '../../config/Firebase';
import { Colors, NUSEmailSignature } from '../../constants';
import { AuthButton, FormInput, ErrorMessage, MainText, Popup } from '../../components';

class ForgetPassword extends Component {
    state = {
        // Details
        nusEmail: '',

        // Control - errors
        emailError: '',
        generalError: '',

        // Control - others
        isLoading: false,
        keyboardShown: false,
        keyboardHeight: 0,
        resetSuccessPopup: false,
    };

    clearInputs() {
        this.setState({
            nusEmail: '',
            emailError: '',
            generalError: '',
        });
    }

    componentDidMount() {
        this.keyboardDidShowListener = Keyboard.addListener(
            'keyboardDidShow',
            this._keyboardDidShow.bind(this)
        );
        this.keyboardDidHideListener = Keyboard.addListener(
            'keyboardDidHide',
            this._keyboardDidHide.bind(this)
        );
    }

    _keyboardDidShow(event) {
        console.log('Keyboard Shown');
        this.setState({
            keyboardShown: true,
            keyboardHeight: event.endCoordinates.height,
        });
    }

    _keyboardDidHide() {
        console.log('Keyboard Hidden');
        this.setState({
            keyboardShown: false,
        });
    }

    componentWillUnmount() {
        this.keyboardDidHideListener.remove();
        this.keyboardDidShowListener.remove();
    }

    onResetSuccess() {
        this.toggleResetSuccessPopup();
    }

    onResetFailure(error) {
        let errorCode = error.code;
        let errorMessage = error.message;

        if (errorCode === 'auth/user-not-found') {
            // Indicates the user account was not found.
            this.setState({
                emailError: 'Unregistered email address',
            });
        } else if (errorCode === 'auth/invalid-email') {
            // 17008: Indicates the email is invalid.
            this.setState({
                emailError: 'Invalid email address',
            });
        } else if (errorCode === 'auth/network-request-failed') {
            // Indicates network error
            this.setState({ generalError: 'Network error' });
        } else {
            // Others
            this.setState({ generalError: 'Unknown error' });
            console.warn('Unknown error: ' + errorCode + ' ' + errorMessage);
        }
    }

    goToSignIn() {
        this.clearInputs.bind(this)();
        this.props.navigation.navigate('SignIn');
    }

    handleEmail(text) {
        this.setState({ nusEmail: text });
    }

    clearError() {
        this.setState({
            emailError: '',
        });
    }

    async resetPassword() {
        const { nusEmail } = this.state;
        this.setState({ isLoading: true });
        try {
            await this.props.firebase.sendPasswordReset(nusEmail);
            console.log('Password reset email sent successfully');
            this.onResetSuccess.bind(this)();
        } catch (error) {
            this.onResetFailure.bind(this)(error);
        } finally {
            this.setState({ isLoading: false });
        }
    }

    validateInput() {
        if (!String(this.state.nusEmail).includes(NUSEmailSignature)) {
            this.setState({ emailError: 'Invalid email domain' });
            return null;
        }
        return this.resetPassword.bind(this)();
    }

    renderResetSuccessPopup = () => {
        return (
            <Popup
                imageType={'Success'}
                isVisible={this.state.resetSuccessPopup}
                title={'Email link sent'}
                body={
                    'We sent an email to\n' +
                    this.state.nusEmail +
                    '\nwith a link to get back into your account.'
                }
                buttonText={'OK'}
                callback={() => {
                    this.toggleResetSuccessPopup();
                    this.setState({
                        nusEmail: '',
                    });
                }}
            />
        );
    };

    toggleResetSuccessPopup = () => {
        this.setState({
            resetSuccessPopup: !this.state.resetSuccessPopup,
        });
    };

    render() {
        const { nusEmail, emailError, generalError, isLoading, keyboardShown } = this.state;

        return (
            <View
                style={[
                    styles.container,
                    {
                        paddingBottom:
                            Platform.OS === 'ios' && keyboardShown
                                ? this.state.keyboardHeight
                                : null,
                    },
                ]}
            >
                {this.renderResetSuccessPopup()}
                <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
                    <View>
                        <View style={styles.textContainer}>
                            <MainText style={styles.title}>Trouble with logging in?</MainText>
                            <MainText style={styles.intro}>
                                Enter your email address and we'll send you a link to reset your
                                password
                            </MainText>
                        </View>

                        <View style={styles.form}>
                            <FormInput
                                containerStyle={styles.box}
                                isError={emailError}
                                errorMessage={emailError}
                                leftIconName="ios-mail"
                                placeholder="NUS email address"
                                returnKeyType="done"
                                keyboardType="email-address"
                                textContentType="emailAddress"
                                autoCapitalize="none"
                                value={nusEmail}
                                onChangeText={this.handleEmail.bind(this)}
                                onFocus={this.clearError.bind(this)}
                            />
                            <View>
                                <AuthButton
                                    onPress={this.validateInput.bind(this)}
                                    style={styles.button}
                                    loading={isLoading}
                                >
                                    Next
                                </AuthButton>
                                <ErrorMessage
                                    error={generalError}
                                    style={{ textAlign: 'center' }}
                                />
                            </View>
                        </View>

                        {keyboardShown ? null : (
                            <View style={styles.bottom}>
                                <MainText
                                    style={styles.backLoginText}
                                    onPress={this.goToSignIn.bind(this)}
                                >
                                    Back to Login
                                </MainText>
                            </View>
                        )}
                    </View>
                </TouchableWithoutFeedback>
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: 'column',
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: Colors.appWhite,
    },
    title: {
        fontSize: 25,
        color: Colors.appGreen,
        textAlign: 'center',
        marginBottom: 10,
    },
    textContainer: {
        flex: 3,
        flexDirection: 'column',
        justifyContent: 'flex-end',
        marginHorizontal: 5,
    },
    intro: {
        fontSize: 15,
        flexWrap: 'wrap',
        textAlign: 'center',
    },
    form: {
        flex: 2,
        justifyContent: 'flex-start',
        marginTop: 30,
        marginHorizontal: 40,
    },
    bottom: {
        flex: 1,
        justifyContent: 'flex-end',
        marginBottom: 36,
        alignItems: 'center',
    },
    box: {
        marginTop: 10,
    },
    button: {
        marginTop: 10,
    },
    backLoginText: {
        position: 'absolute',
        bottom: 0,
        fontWeight: '200',
        fontSize: 15,
        textAlign: 'center',
        color: Colors.appGreen,
        marginBottom: 20,
    },
});

export default withFirebase(ForgetPassword);
