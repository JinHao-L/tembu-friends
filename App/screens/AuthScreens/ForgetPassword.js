import React, { Component } from 'react';
import {
    View,
    StyleSheet,
    KeyboardAvoidingView,
    Platform,
    TouchableWithoutFeedback,
    Keyboard,
} from 'react-native';

import { withFirebase } from '../../config/Firebase';
import { Colors, NUSEmailSignature } from '../../constants';
import { AuthButton, FormInput, ErrorMessage, MainText, textStyles } from '../../components';

class ForgetPassword extends Component {
    state = {
        // Details
        nusEmail: '',

        // Control - errors
        emailError: '',
        generalError: '',

        // Control - others
        isLoading: false,
        isSuccess: false,
    };

    onResetSuccess() {
        this.setState({
            nusEmail: '',
            isSuccess: true,
        });
        //TODO: Pop-up here
        // this.props.navigation.navigate('SignIn');
    }

    onResetFailure(error) {
        let errorCode = error.code;
        let errorMessage = error.message;

        if (errorCode === 'auth/user-not-found') {
            // Indicates the user account was not found.
            this.setState({
                emailError: 'Unregistered email address',
            });
        } else if (errorCode === 'auth/invalid-email') {
            // 17008: Indicates the email is invalid.
            this.setState({
                emailError: 'Invalid email address',
            });
        } else if (errorCode === 'auth/network-request-failed') {
            // Indicates network error
            this.setState({ generalError: 'Network error' });
        } else {
            // Others
            this.setState({ generalError: 'Unknown error' });
            console.warn('Unknown error: ' + errorCode + ' ' + errorMessage);
        }
    }

    goToSignIn() {
        this.props.navigation.navigate('SignIn');
    }

    handleEmail(text) {
        this.setState({ nusEmail: text, emailError: '' });
    }

    async resetPassword() {
        const { nusEmail } = this.state;
        this.setState({ isLoading: true });
        try {
            await this.props.firebase.passwordReset(nusEmail);
            console.log('Password reset email sent successfully');
            this.onResetSuccess.bind(this)();
        } catch (error) {
            this.onResetFailure.bind(this)(error);
        } finally {
            this.setState({ isLoading: false });
        }
    }

    validateInput() {
        if (!String(this.state.nusEmail).includes(NUSEmailSignature)) {
            this.setState({ emailError: 'Invalid email domain' });
            return null;
        }
        return this.resetPassword.bind(this)();
    }

    render() {
        const { nusEmail, emailError, generalError, isLoading, isSuccess } = this.state;

        return (
            <KeyboardAvoidingView
                behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
                style={styles.container}
                keyboardVerticalOffset={Platform.OS === 'ios' ? 40 : -150}
                contentContainerStyle={{ flex: 1 }}
            >
                <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
                    <View>
                        <View style={styles.textContainer}>
                            <MainText style={styles.title}>Trouble with logging in?</MainText>
                            <MainText style={styles.intro}>
                                Enter your email address and we'll send you a link to reset your
                                password
                            </MainText>
                        </View>

                        <View style={styles.form}>
                            <View style={styles.box}>
                                <FormInput
                                    style={emailError ? styles.errorInput : styles.validInput}
                                    leftIconName="ios-mail"
                                    placeholder="NUS email address"
                                    returnKeyType="done"
                                    keyboardType="email-address"
                                    textContentType="emailAddress"
                                    autoCapitalize="none"
                                    value={nusEmail}
                                    onChangeText={this.handleEmail.bind(this)}
                                />
                                <ErrorMessage error={emailError ? emailError : ' '} />
                            </View>
                            <View>
                                <AuthButton
                                    onPress={this.validateInput.bind(this)}
                                    style={[styles.button, { position: 'relative', top: 10 }]}
                                    loading={isLoading}
                                >
                                    Next
                                </AuthButton>
                                <ErrorMessage error={generalError} />
                                <MainText style={styles.successMessage}>
                                    {isSuccess ? 'Reset mail successfully sent' : null}
                                </MainText>
                            </View>
                        </View>

                        <View style={styles.bottom}>
                            <MainText
                                style={styles.backLoginText}
                                onPress={this.goToSignIn.bind(this)}
                            >
                                Back to Login
                            </MainText>
                        </View>
                    </View>
                </TouchableWithoutFeedback>
            </KeyboardAvoidingView>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: 'column',
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: Colors.authBackground,
    },
    title: {
        fontSize: 30,
        // fontWeight: '700',
        color: Colors.greenText,
        marginBottom: 10,
        textAlign: 'center',
    },
    textContainer: {
        flex: 3,
        flexDirection: 'column',
        justifyContent: 'flex-end',
    },
    intro: {
        flexWrap: 'wrap',
        textAlign: 'center',
    },
    form: {
        flex: 2,
        justifyContent: 'flex-start',
        marginTop: 30,
        marginHorizontal: 30,
    },
    bottom: {
        flex: 1,
        justifyContent: 'flex-end',
        marginBottom: 36,
        alignItems: 'center',
    },
    validInput: {
        borderColor: Colors.defaultBorder,
    },
    errorInput: {
        borderColor: Colors.errorBorder,
    },
    box: {
        marginTop: 10,
    },
    button: {
        marginTop: 10,
    },
    successMessage: {
        color: Colors.successText,
        textAlign: 'center',
        fontSize: 16,
    },
    backLoginText: {
        position: 'absolute',
        bottom: 0,
        fontWeight: '200',
        fontSize: 15,
        textAlign: 'center',
        color: Colors.greenText,
        marginBottom: 20,
    },
});

export default withFirebase(ForgetPassword);
