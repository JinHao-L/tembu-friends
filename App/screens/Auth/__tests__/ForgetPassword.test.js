import 'react-native';
import React from 'react';
import ForgetPassword from '../ForgetPassword';

import renderer from 'react-test-renderer';

describe('<ForgetPassword/>', () => {
    const tree = renderer.create(<ForgetPassword />).toJSON();

    it('renders correctly', () => {
        expect(tree).toMatchSnapshot();
    });

    it('has 1 child', () => {
        expect(tree.children.length).toBe(1);
    });
});
