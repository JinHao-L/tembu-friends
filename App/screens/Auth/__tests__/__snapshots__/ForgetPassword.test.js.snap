// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgetPassword/> renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "flex": 1,
      },
      Object {
        "paddingBottom": null,
      },
    ]
  }
>
  <Modal
    animated={true}
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    theme={
      Object {
        "colors": Object {
          "disabled": "hsl(208, 8%, 90%)",
          "divider": "#bcbbc1",
          "error": "#ff190c",
          "grey0": "#393e42",
          "grey1": "#43484d",
          "grey2": "#5e6977",
          "grey3": "#86939e",
          "grey4": "#bdc6cf",
          "grey5": "#e1e8ee",
          "greyOutline": "#bbb",
          "platform": Object {
            "android": Object {
              "error": "#f44336",
              "primary": "#2196f3",
              "secondary": "#9C27B0",
              "success": "#4caf50",
              "warning": "#ffeb3b",
            },
            "ios": Object {
              "error": "#ff3b30",
              "primary": "#007aff",
              "secondary": "#5856d6",
              "success": "#4cd964",
              "warning": "#ffcc00",
            },
          },
          "primary": "#2089dc",
          "searchBg": "#303337",
          "secondary": "#8F0CE8",
          "success": "#52c41a",
          "warning": "#faad14",
        },
      }
    }
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, .4)",
          "height": "100%",
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
      testID="RNE__Overlay__backdrop"
    />
    <View
      pointerEvents="box-none"
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 20,
            "marginHorizontal": 0,
            "padding": 10,
            "paddingHorizontal": 0,
            "paddingVertical": 0,
            "shadowColor": "rgba(0, 0, 0, .3)",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowRadius": 4,
            "width": 375,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "width": "100%",
            }
          }
        >
          <View>
            <Image
              resizeMode="contain"
              source={1}
              style={
                Object {
                  "alignSelf": "center",
                  "height": 80,
                  "marginTop": 5,
                  "top": 10,
                  "width": 150,
                }
              }
            />
            <View
              style={
                Object {
                  "alignItems": "center",
                  "paddingHorizontal": 30,
                  "paddingVertical": 20,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Montserrat-SemiBold",
                      "fontSize": 13,
                    },
                    Object {
                      "color": "#333",
                      "fontSize": 18,
                      "fontWeight": "600",
                      "justifyContent": "center",
                      "paddingVertical": 10,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Email link sent
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Montserrat-SemiBold",
                      "fontSize": 13,
                    },
                    Object {
                      "color": "#666",
                      "fontSize": 15,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                We sent an email to

with a link to get back into your account.
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "borderBottomColor": "#F1F1F1",
                "borderWidth": 0.5,
                "width": 375,
              }
            }
          />
          <View
            style={
              Object {
                "borderBottomEndRadius": 20,
                "borderBottomStartRadius": 20,
                "borderRadius": 0,
                "overflow": "hidden",
                "width": 375,
              }
            }
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "busy": false,
                  "disabled": false,
                }
              }
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "borderColor": "#2089dc",
                    "borderRadius": 3,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "padding": 8,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#222",
                      "fontFamily": "Montserrat-SemiBold",
                      "fontSize": 18,
                      "fontWeight": "600",
                      "paddingBottom": 1,
                      "paddingTop": 2,
                      "textAlign": "center",
                    }
                  }
                >
                  OK
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </Modal>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "justifyContent": "center",
        },
        Object {
          "marginTop": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginHorizontal": 15,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Montserrat-SemiBold",
              "fontSize": 13,
            },
            Object {
              "color": "#248458",
              "fontSize": 25,
              "marginBottom": 10,
              "textAlign": "center",
            },
          ]
        }
      >
        Trouble with logging in?
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Montserrat-SemiBold",
              "fontSize": 13,
            },
            Object {
              "flexWrap": "wrap",
              "fontSize": 15,
              "textAlign": "center",
            },
          ]
        }
      >
        Enter your email address and we'll send you a link to reset your password
      </Text>
    </View>
    <View
      style={
        Object {
          "marginHorizontal": 40,
          "marginTop": 30,
        }
      }
    >
      <View
        style={
          Array [
            undefined,
            Object {
              "marginTop": 10,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "borderRadius": 6,
                "borderWidth": 1,
                "flexDirection": "row",
              },
              undefined,
              Object {
                "borderColor": "#D9D9D9",
              },
            ]
          }
        >
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            keyboardType="email-address"
            leftIconName="ios-mail"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="NUS email address"
            placeholderTextColor="#D9D9D9"
            rejectResponderTermination={true}
            returnKeyType="done"
            style={
              Object {
                "flex": 1,
                "fontFamily": "Montserrat-SemiBold",
                "fontSize": 15,
                "fontWeight": "600",
                "marginLeft": 5,
                "paddingVertical": 5,
                "textAlignVertical": "center",
              }
            }
            textContentType="emailAddress"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Montserrat-SemiBold",
                "fontSize": 13,
              },
              Array [
                Object {
                  "color": "#FF0000",
                  "flexWrap": "wrap",
                  "fontSize": 12,
                  "textAlign": "left",
                },
                undefined,
              ],
            ]
          }
        >
           
        </Text>
      </View>
      <View>
        <RNGestureHandlerButton
          activeOpacity={0.105}
          collapsable={false}
          onActiveStateChange={[Function]}
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          onPress={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Array [
                Object {
                  "borderRadius": 5,
                  "paddingHorizontal": 5,
                  "paddingVertical": 7,
                },
                Object {
                  "backgroundColor": "#248458",
                },
                Object {
                  "marginTop": 10,
                },
              ],
            ]
          }
          underlayColor="black"
        >
          <View
            style={
              Object {
                "backgroundColor": "black",
                "bottom": 0,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Montserrat-SemiBold",
                  "fontSize": 13,
                },
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 15,
                  "fontWeight": "600",
                  "textAlign": "center",
                },
              ]
            }
          >
             
            Next
             
          </Text>
        </RNGestureHandlerButton>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Montserrat-SemiBold",
                "fontSize": 13,
              },
              Array [
                Object {
                  "color": "#FF0000",
                  "flexWrap": "wrap",
                  "fontSize": 12,
                  "textAlign": "left",
                },
                Object {
                  "textAlign": "center",
                },
              ],
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "marginBottom": 56,
        "position": "absolute",
      }
    }
  >
    <Text
      onPress={[Function]}
      style={
        Array [
          Object {
            "fontFamily": "Montserrat-SemiBold",
            "fontSize": 13,
          },
          Object {
            "color": "#248458",
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    >
      Back to Login
    </Text>
  </View>
</View>
`;
